packages/repository/src/query.ts

```diff
@@ -194,91 +265,120 @@ export class WhereBuilder {
     return this;
   }
 
+  /**
+   * Cast an `and` or `or` clause to Where
+   * @param clause And/Or clause
+   */
+  cast(clause: AndClause<MT> | OrClause<MT>) {
+    const w: Where<MT> = {};
+    if ('and' in clause) {
+      w.and = clause.and;
+    }
+    if ('or' in clause) {
+      w.or = clause.or;
+    }
+    return w;
+  }
+
   /**
    * Add an `and` clause.
    * @param w One or more where objects
    */
-  and(...w: (Where | Where[])[]): this {
-    let clauses: Where[] = [];
+  and(...w: (Where<MT> | Where<MT>[])[]): this {
+    let clauses: Where<MT>[] = [];
     w.forEach(where => {
       clauses = clauses.concat(Array.isArray(where) ? where : [where]);
     });
-    return this.add({and: clauses});
+    return this.add(this.cast({and: clauses}));
   }
 
   /**
    * Add an `or` clause.
    * @param w One or more where objects
    */
-  or(...w: (Where | Where[])[]): this {
-    let clauses: Where[] = [];
+  or(...w: (Where<MT> | Where<MT>[])[]): this {
+    let clauses: Where<MT>[] = [];
     w.forEach(where => {
       clauses = clauses.concat(Array.isArray(where) ? where : [where]);
     });
-    return this.add({or: clauses});
+    return this.add(this.cast({or: clauses}));
   }
 
   /**
    * Add an `=` condition
    * @param key Property name
    * @param val Property value
    */
-  eq(key: string, val: any): this {
-    return this.add({[key]: val});
+  eq<K extends KeyOf<MT>>(key: K, val: MT[K]): this {
+    const w: Where<MT> = {};
+    w[key] = val;
```

Comment:
I think this can be simplified as follows - see the original code which was already using this simpler form!

```ts
return this.add(w, {[key]: val});
```

The same comment applies to builder methods below too.